x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/SecurityContext.java

```diff
@@ -149,8 +157,27 @@ public void executeAfterRewritingAuthentication(Consumer<StoredContext> consumer
         final Authentication authentication = getAuthentication();
         try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
             setAuthentication(new Authentication(authentication.getUser(), authentication.getAuthenticatedBy(),
-                authentication.getLookedUpBy(), version, authentication.getAuthenticationType(), authentication.getMetadata()));
+                authentication.getLookedUpBy(), version, authentication.getAuthenticationType(),
+                rewriteMetadataIfNecessary(version, authentication)));
             consumer.accept(original);
         }
     }
+
+    private Map<String, Object> rewriteMetadataIfNecessary(Version streamVersion, Authentication authentication) {
+        Map<String, Object> metadata = authentication.getMetadata();
+        if (authentication.getAuthenticationType() == AuthenticationType.API_KEY
+            && authentication.getVersion().onOrAfter(Version.V_7_9_0)
+            && streamVersion.before(Version.V_7_9_0)) {
```

Comment:
These are all good suggestions. Will address accordingly.

Reply:
I renamed the method to `rewriteMetadataForApiKeyRoleDescriptors`. But after some thoughts, I decided to keep this method inside `SecurityContext`. Because I need to make it public if moved to `Authentication` and that feels a bit uneasy to me since `SecurityContext` is the only consumer and will probably keep it that way in future. So I feel it is better to stay within it and closer to the actual context. It is an unnecessary detail to put inside `Authentication` if people is not concerned of BWC. Similarly I added the new version constant to `SecurityContext` as well. Sorry for the back and forth.

Reply:
Okay, but the constant can sit inside `Authentication` to be referenced in the `CompositeRolesStore` too.

Reply:
Moved the new version constant to `Authentication` as suggested.